src/predictor/array_tree_layout.h

```diff
@@ -0,0 +1,300 @@
+/**
+ * Copyright 2021-2025, XGBoost Contributors
+ * \file array_tree_layout.cc
+ * \brief Implementation of array tree layout -- a powerfull inference optimization method.
+ */
+#ifndef XGBOOST_PREDICTOR_ARRAY_TREE_LAYOUT_H_
+#define XGBOOST_PREDICTOR_ARRAY_TREE_LAYOUT_H_
+
+#include <limits>
+#include <vector>
+
+#include "predict_fn.h"             // for GetNextNode, GetNextNodeMulti
+#include "xgboost/tree_model.h"     // for RegTree, MTNotImplemented, RTreeNodeStat
+
+namespace xgboost::predictor {
+
+/**
+ * @brief The class holds the array-based representation of the top levels of a single tree.
+ * 
+ * @tparam TreeType The type of the origianl tree (RegTree or MultiTargetTree)
+ *
+ * @tparam has_categorical if the tree has categorical features
+ *
+ * @tparam any_missing if the class is able to process missing values
+ * 
+ * @tparam kNumDeepLevels number of tree leveles being unrolled into array-based structure
+ */
+template <class TreeType, bool has_categorical, bool any_missing, int kNumDeepLevels>
+class ArrayTreeLayout {
+ private:
+  /* Number of nodes in the array based representation of the top levels of the tree
+   */
+  constexpr static size_t kNodesCount = (1u << kNumDeepLevels) - 1;
+
+  struct Empty {};
+  using DefaultLeftType =
+        typename std::conditional_t<any_missing,
+                                   std::array<uint8_t, kNodesCount>,
+                                   struct Empty>;
+  using IsCatType =
+        typename std::conditional_t<has_categorical,
+                                   std::array<uint8_t, kNodesCount>,
+                                   struct Empty>;
+  using CatSegmentType =
+        typename std::conditional_t<has_categorical,
+                                   std::array<common::Span<uint32_t const>, kNodesCount>,
+                                   struct Empty>;
+
+  DefaultLeftType default_left_;
+  IsCatType is_cat_;
+  CatSegmentType cat_segment_;
+
+  std::array<bst_feature_t, kNodesCount> split_index_;
+  std::array<float, kNodesCount> split_cond_;
+  /* The nodes at tree levels 0, 1, ..., kNumDeepLevels - 1 are unrolled into an array-based structure.
+   *  If the tree has additional levels, this array stores the node indices of the sub-trees at level kNumDeepLevels.
+   *  This is necessary to continue processing nodes that are not eligible for array-based unrolling.
+   *  The number of sub-trees packed into this array is equal to the number of nodes at tree level kNumDeepLevels,
+   *  which is calculated as (1u << kNumDeepLevels) == kNodesCount + 1.
```

Comment:
What happens if the tree is not well balanced and is more like a linked list than a tree?

Reply:
In this case we add dummy nodes with `nan` as split condition. In this dummy node the decision is always "go right" (see also comments https://github.com/razdoburdin/xgboost/blob/b0eaa856e1246416f7f9538bcc004e7723d9b997/src/predictor/array_tree_layout.h#L154), the left child are not initialized. 

So with array layout we have to allocated all nodes, but keep some of the unpopulated in case the tree is pure balanced.

Initial tree:
<img width="601" height="516" alt="image" src="https://github.com/user-attachments/assets/f23f17f9-3f7a-4119-ba7a-0df1adac3bcd" />

Tree with dummy (nan-valued nodes):
<img width="1287" height="760" alt="image" src="https://github.com/user-attachments/assets/b9bc73a5-6eca-4bc4-ab05-b915f8052cdf" />